Index: rx-codelab/src/test/java/com/jraska/rx/codelab/Task4_ErrorHandling.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- rx-codelab/src/test/java/com/jraska/rx/codelab/Task4_ErrorHandling.java	(date 1490096109000)
+++ rx-codelab/src/test/java/com/jraska/rx/codelab/Task4_ErrorHandling.java	(revision )
@@ -22,23 +22,22 @@
 
     @Test
     public void printErrorMessage() {
-        // TODO: Print all values and incoming error message
+        screwsObservable.subscribe(System.out::println, throwable -> System.out.println(throwable.getMessage()));
     }
 
     @Test
     public void emitCustomItemOnError() {
-        // TODO: When an error happens, emit number artificial screw
+        screwsObservable.onErrorReturnItem(Parts.screw()).subscribe(System.out::println);
     }
 
     @Test
     public void subscribeToExtraObservableOnError() {
-        // TODO: When an error happens, subscribe to extra observable
+        screwsObservable.onErrorResumeNext(extraScrewsObservable).subscribe(System.out::println);
     }
 
     @Test
     public void retryOnError() {
-        Observable<Screw> flakeyObservable = Parts.flakeyScrew();
-        // TODO: Observable is a bit flakey and often fails, use retry to make it always complete
+        Observable<Screw> flakeyObservable = Parts.flakeyScrew().retry();
 
         for (int i = 0; i < 10; i++) {
             flakeyObservable.subscribe(System.out::println);
